fastlane/fastlane
# [ci] add GitHub Action workflow to label PRs based on tools they touch

### Checklist


### Motivation and Context

We keep adding these labels manually, but only when a core contributor opens/reviews a PR. A handy tool that does this automatically for us, even before we review the PR, would be helpful!

### Description

This workflow will automatically label PRs with e.g. `tool: fastlane_core`, `type: ci`, `tool: deliver`, etc. based on the directories that are being modified in the PR.

I'm wondering if I forgot any obvious rule we should be adding? What are other files associated to the `type: ci` label? Are there other labels we could automate based on file patterns?

> [!IMPORTANT]
> This automation will only kick in when PRs are open/reopened or a new commit is pushed to it. In other words, it won't update existing open PRs. We'd need to do this in a one-off basis somehow. I intend to rebase all open PRs so that should help adding labels to all PRs that don't have merge conflicts currently, otherwise, we could merge a GHA workflow that will be trigger on cron job and then remove it once it's run once. Lmk your thoughts on this 😇 

### Testing Steps

Check if it adds and removes the labels as promised, in this PR. Then, switch `pull_request` trigger to `pull_request_target`, so it works with forked PRs. Read this (https://github.com/actions/labeler#notes-regarding-pull_request_target-event) to understand why this switching is needed to test.

#### Screenshots of tests observed in this PR:

Labels successfully added by the bot once everything was configured correctly:

<img width="910" alt="image" src="https://github.com/fastlane/fastlane/assets/8419048/e4d896ab-94dd-428d-9845-7d5580020e8b">
<img width="333" alt="image" src="https://github.com/fastlane/fastlane/assets/8419048/73263276-4996-41e3-ba91-a019a40d7098">

Labels successfully removed when the code changed in those directories was removed:

<img width="810" alt="image" src="https://github.com/fastlane/fastlane/assets/8419048/98a254b2-ed8f-41bb-90db-2b7bf0e17081">

Not sure if we should allow the bot to remove labels, see this comment of mine: https://github.com/fastlane/fastlane/pull/21751#discussion_r1435292843
